hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+ *   findFixups - Finds the affected fixups following a relocation.
+ *   relocate - Relocates a range of instructions to a new
+ *              destination. Critically, relocation is a chance
+ *              to grow/shrink the region which is advantageous
+ *              when moving a far branch to simple near relative
+ *              branch for instance.
+ *
+ * adjustForRelocation and adjustCodeForRelocation both search for
+ * addresses in instruction ranges that need to be updated following
+ * a relocation. This involves finding instructions which might
+ * be loading an address and then determining if that target is
+ * actually an address for an instruction which has been moved. For
+ * ARM, this includes the following:
+ *
+ *   PC Relative
+ *     ADR/ADRP - Builds PC relatative addresses.
+ *     B[.<cc>] (immediate) - Branch to a PC relative address.
+ *     LDR (literal) - Loads a literal from a PC relative address.
+ *
+ *   Immediates
+ *     MOV/MOK - Used to build up targets 16 bits at a time.
+ *     LDR (literal) - The literal loaded may be a target address.
+ *
+ * adjustCodeForRelocation must support updating live instructions.
+ * For ARM, we can update a single instruction or a 64 bit literal
+ * in the instruction stream atomically. We can't, however, update
+ * multiple instructions atomically since there may be a thread
+ * executing instructions within that range. Therefore, MOV/MOVK
+ * instructions can't be adjusted for live ranges of instructions.
+ * Any emitted code that uses MOV/MOVK and which must be adjusted
+ * while live, should instead use LDR (literal). Smashables use this
+ * approach; they embed a literal in the instruction stream which
+ * can then be updated atomically. At the time of this writing, only
+ * the vasm opcode (leap) produces a live region that must be updated,
+ * and so it uses an approach with LDR (literal) instead of MOV/MOVK.
+ */
+
+
+//////////////////////////////////////////////////////////////////////
+
+using InstrSet = hphp_hash_set<Instruction*>;
+struct JmpOutOfRange : std::exception {};
+
+TcaRange fixupRange(const RelocationInfo& rel, const TcaRange& rng) {
+  /*
+   * We have to be careful with before/after here.
+   * If we relocate two consecutive regions of memory,
+   * but relocate them to two different destinations, then
+   * the end address of the first region is also the start
+   * address of the second region; so adjustedAddressBefore(end)
+   * gives us the relocated address of the end of the first
+   * region, while adjustedAddressAfter(end) gives us the
+   * relocated address of the start of the second region.
+   */
+  auto s = rel.adjustedAddressAfter(rng.begin());
+  auto e = rel.adjustedAddressBefore(rng.end());
+  if (s && e) {
+    return TcaRange(s, e);
+  }
+  if (s && !e) {
+    return TcaRange(s, s + rng.size());
+  }
+  if (!s && e) {
+    return TcaRange(e - rng.size(), e);
+  }
+  return rng;
+}
```

Comment:
This looks like it's duplicated for every architecture.  Can you just add it to `relocation.h` and implement it a single time in `relocation.cpp`?